### PR TITLE
Fix segfault caused by `ThreadStatus`

### LINKED ISSUE #51926: Segfault during insert in `01527_materialized_view_stack_overflow` (`ProfileEvents::Counters::incrementNoTrace`)
https://s3.amazonaws.com/clickhouse-test-reports/0/2443ccdcf6be98d5ff8e74807f9988666754f82d/stateless_tests__debug__%5B3_5%5D.html

In `gdb.log`: https://pastila.nl/?000cb995/6ead8a43d668cf35b447f0bae4562127

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Better check for `current_thread` in `ThreadStatus` to avoid segfault.

Fix https://github.com/ClickHouse/ClickHouse/issues/51926

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by Avogar — APPROVED** [review]

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 408e409f7ccaab334ac4be5bfaacf8ccb92bbe34 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51931/408e409f7ccaab334ac4be5bfaacf8ccb92bbe34/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Common/ThreadStatus.cpp [hunk 1]
```diff
@@ -199,13 +199,12 @@ ThreadStatus::~ThreadStatus()
     if (deleter)
         deleter();
 
+    assert(!check_current_thread_on_destruction || current_thread == this);
```
Comment (Avogar) [inline src/Common/ThreadStatus.cpp:202]: Maybe we can also add `LOG_ERROR` if `!check_current_thread_on_destruction || current_thread == this` to track this in release build. I am trying to debug https://github.com/ClickHouse/ClickHouse/issues/51780 but such failures occures only in release builds (maybe with sanitizers)

Comment (antonio2368) [reply src/Common/ThreadStatus.cpp:202]: throw `LOGICAL_ERROR` maybe?

Comment (Avogar) [reply src/Common/ThreadStatus.cpp:202]: From destructor? Then we will have possible aborts in release build

Comment (antonio2368) [reply src/Common/ThreadStatus.cpp:202]: yeah, but I saw `adjustWithUntrackedMemory` so thought maybe there is no difference
will add `LOG_ERROR`

**Review by Avogar — APPROVED** [review]

**Pull request merged**